textx/metamodel.py

```diff
@@ -485,7 +494,16 @@ def validate(self):
         Validates metamodel. Called after construction to check for some
         textX rules.
         """
-        # TODO: Implement complex textX validations.
+        from textx.exceptions import TextXSemanticError
+        for user_class in self.user_classes.values():
+            if not hasattr(user_class, "_tx_obj_attrs"):
+                # It is not a user class used in the grammar
+                # Note: see textx.lang.visit_rule_name, where
+                #    metamodel._init_class is called with
+                #    external_attributes=True.
+                raise TextXSemanticError(
+                    "unexpected: {} seems to be unused in the grammar".format(
+                        user_class.__name__))
```

Comment:
Not completely sure about this check, but lets go with it for now. I would just change the wording of the exception to "{} class is not used in the grammar".

Reply:
ok

Reply:
Now we have a separate function to validate user classes. This new validation function is only called for the meta model for the main grammar (not for imported grammars).

Reply:
Now, the check has been clarified: we explicitly check for used rule names...